mx.py

```diff
@@ -8714,19 +8718,25 @@ def __init__(self, suite, name, path, optional, urls, sha1, sourcePath, sourceUr
         BaseLibrary.__init__(self, suite, name, optional, theLicense, **kwArgs)
         ClasspathDependency.__init__(self, **kwArgs)
         self.path = path.replace('/', os.sep) if path is not None else None
-        self.urls = [self.substVars(url) for url in urls]
-        self.sha1 = mx_urlrewrites.rewritesha1(self.urls, sha1)
+        self.urls = urls
```

Comment:
Is there any point after this constructor where we need to access the original `urls` or the `urlsSubstituted`? It seems safer if we do not let these intermediate value escape outside the constructor so we should simply store the substituted and rewritten value in `self.urls` (same for `self.sourceSha1`).

Reply:
I'll fix this as part of integrating the PR.

Reply:
The urls are likely not needed, however, it might be useful to keep them in `self.urlsRewritten` rather than just `self.urls` to avoid leftover code using the rewritten values by mistake. The sha1 needs to stay in both original and rewritten form otherwise`_get_path_in_cache` will fail.

Thanks for looking into this !

Reply:
> The sha1 needs to stay in both original and rewritten form otherwise_get_path_in_cache will fail.

I'm not following. Is there a manual set of steps I can perform that will show me the problem?

Reply:
To demonstrate, you would need to do an SHA1 rewrite on a `PackedResourceLibrary` resource. We do that in our infra for example with Renaissance 0.13 AOT runs, with this:

```
export MX_URLREWRITES='[
    {"https://lafo.ssw.uni-linz.ac.at/pub/graal-external-deps/renaissance/renaissance-gpl-0.13.0.jar":
        {"replacement":"file:./../renaissance.jar","sha1":"5ee38f52755f2544e23a2d1bb9c86c3f0742915b"}},
    {"https://lafo.ssw.uni-linz.ac.at/pub/graal-external-deps/renaissance/renaissance-harness_v0.13.0.tar.gz":
        {"replacement":"file:./../harness.tar.gz","sha1":"7029d2211235185ecd492b6092b95dbbaa5171d5"}}
]'
```

What happens is, mx will call `_get_path_in_cache` from `_load_libraries` to create the `path` argument for `PackedResourceLibrary` constructor. That path will include the original SHA1. Inside the constructor, SHA1 is rewritten and the path provided as the argument is used as `extract_path`, while `path` is computed again, this time with rewritten SHA1, through another call to `_get_path_in_cache`. This inconsistency (`path` and `extract_path` based on different values of SHA1) can later lead to crash in other mx code, in our case it was `mx_substratevm_benchmark.py` from Graal calling `get_path`, which in turn calls `SafeDirectoryUpdater`, which fails when trying to create a temporary directory because the parent does not exist.

Reply:
Sorry for dragging this out but I'd really prefer not to do a half fix now and have to revisit this again in future to clean up the inconsistencies you mention. The high order point is that all substitution and rewriting of URLs and SHA1s should be done in one or as few places as possible. I think that's what @zapster refactoring suggestions were aiming at. In terms of testing, if you do the refactoring and it passes a full internal Graal gate (which I will definitely run), then I believe we're ok.
I assume that you must be using a patched mx locally to solve your immediate problem so I don't think there's an urgency to get a half fix in.

Reply:
Next try :-) here is an update that moves all rewriting inside resource classes, and simplifies `_load_libraries`, as suggested by @zapster. There are only two places in the code now where resource URL rewriting happens (constructors of `Library` and `ResourceLibrary`), the original URL and SHA1 values are no longer kept around.

The code should be tested thoroughly - it works for me with Renaissance and AOT benchmarks, and it appears to work the same as current mx HEAD when building Graal CE, but there are features in the code that are not used by Graal CE.

WRT urgency, we have to use the most current mx with our AOT measurements, without some variant of this fix we do not run some AOT measurements (those that require alternative Renaissance harness to work around Scala library version mismatches). Eventually, these issues may go away due to planned changes to Renaissance, but if we could get some variant of this fix, that would certainly free our hands to focus on other issues.